csspanels@dr.drummie: Add new extension

drdrummie opened the pull request:
csspanels@dr.drummie: Add new extension

A new Cinnamon extension that customizes the appearance of panels, popups, banners, and OSD elements using CSS blur effects.

This extension allows users to apply modern, blurred "glass" UI styles to various Cinnamon desktop components, offering enhanced personalization options.  
While partially based on existing code, it introduces different functionality and additional customization features.

All usage instructions are provided in the README.md.

drdrummie committed:
Add author information to info.json

drdrummie committed:
Add files via upload

drdrummie committed:
Add metadata.json for CSS Panels extension

drdrummie committed:
Revise credits in README.md

Updated credits section to include icon source.

drdrummie committed:
Added extension files

drdrummie committed:
Update compatibility section in README.md to reflect supported Cinnamon versions

drdrummie committed:
Update metadata and translation files for CSS Panels extension

drdrummie committed:
Update icon attribution in README.md and replace icon image

drdrummie committed:
Fix typo in README.md regarding CSS blur templates

drdrummie committed:
Enhance troubleshooting section in README

Updated the troubleshooting section to include known issues and additional guidance.

drdrummie committed:
New version of the extension 1.8.9.:
- Updated the TransparencyManager class to handle and style multiple panels (style of "main" panel is propagated to other panels as well if they exist).
- Updated README.md.
- Added CHANGELOG.md.
- Updated translations (pot / po files).

drdrummie committed:
New version of the extension 1.9.2.:
- Added styling for tooltips and app switchers.
- Fixed known bugs and refactored most of the code.
- Updated README.md. and CHANGELOG.md.
- Updated translations (pot / po files).
- Improved logging and troubleshooting.

fredcw commented:
@drdrummie Ah sorry, the warnings are nothing to do with your extension, I always get those the first time I open a context menu or app menu after restarting cinnamon.

I think that most people would install the extension specifically for the blur. If it doesn't have that then I think most people would be somewhat disappointed given what it says in the readme and description.

drdrummie committed:
Update README and metadata for clarity on visual effects terminology

fredcw commented:
@drdrummie tbh, given that not all features work on most themes (especially the Mint themes), it might be frustrating to users unless it's specified which theme's it does work on. Another thing to consider is that many people won't read the readme before installing it through cinnamon's UI. This is just my opinion of course. I think it may have potential if it could be made to work reliably on most or nearly all themes.

drdrummie committed:
v2.0.3 - Major update from v1.9.2 with new features, architectural refactor,
and Cinnamon Spices CI compliance fixes.

New features:
- Wallpaper color extraction (GdkPixbuf) with full-auto and manual modes
- Hover & active color override for panel applets and taskbar items
- Glow Effect Mode: inset / outset / none (replaces border-width approach)
- Desktop context menu styling
- Settings reorganized into 4 pages: Theme, Appearance, Visual Effects, Advanced

Architecture:
- 6 new modules: constants, signalHandler, themeUtils, colorPalette,
  wallpaperMonitor, hoverStyleManager
- Template-based CSS generation (blurTemplateManager)
- Error boundary pattern, LRU cache, debounced theme detection

CI/stability fixes:
- Removed forbidden metadata fields
- Monkey patch idempotency (disable() safe to call multiple times)
- OSD missing disable() fix (leaked monkey patch)
- settings.finalize() in disable()
- Removed untranslated .po files; Croatian (hr) translation maintained

drdrummie committed:
Remove desktop context menu styling feature from README

drdrummie committed:
v2.0.7 - Desklet styling, start menu sidebar, wallpaper extraction mode, dark/light override

- Added desklet styling (transparency, blur, glow for desktop widgets)
- Added start menu sidebar styling (menu@cinnamon.org, disabled by default)
- Added dark/light mode override control for mixed-theme setups
- Added wallpaper extraction mode: Standard (weighted average) vs Contrast (polar tones)
- Added url field back to metadata.json (Cinnamon runtime requirement)
- Updated translations (pot/po files), added en.po
- Updated README: compatibility note, icon attribution, extension conflict warning
- Updated CHANGELOG

drdrummie committed:
v2.0.8: fixes and improvements

- themeDetector: three-state cache prevents repeated sync GTK CSS file reads on the blurTemplateManager hot path
- extension.js, stylerBase.js: fix 2.0.7 regression where debug messages  were routed to logError/logWarning instead of global.log (info level)
- panelStyler: skipThemeRefresh flag avoids redundant _changeTheme() call during re-apply cycles

drdrummie committed:
v2.0.9 - System tray indicator: direct settings link and HoverStyleManager hover integration

- systemIndicator: left click now opens extension settings directly via xlet-settings
  (was: generic cinnamon-settings extensions manager)
- systemIndicator: removed manual enter-event/leave-event opacity handlers; indicator
  registered with HoverStyleManager on create, unregistered on destroy
- hoverStyleManager: added hookExternalActor(actor) / unhookExternalActor(actor) public
  API; external actors tracked in _externalActors[] and auto re-hooked on refresh()
- Updated CHANGELOG and README

drdrummie committed:
Update compatibility: restrict to tested Cinnamon versions (6.4, 6.6)

fredcw commented:
Here's what Gemini thinks. I'm not saying I agree or disagree with any of this, I'm just posting it in case it contains any useful info:

> This is a sophisticated and high-quality Pull Request. It is significantly more advanced than the average Cinnamon extension, resembling the architecture of complex GNOME extensions like *Dash to Panel* or *Blur my Shell*.
> 
> Here is a technical evaluation of the code and my recommendation for its inclusion in the `cinnamon-spices-extensions` repository.
> 
> ---
> 
> ### 1. Technical Evaluation
> 
> #### **Pros (The Strengths)**
> *   **Architecture:** The code is exceptionally modular. The use of a `StylerBase` class and a centralized `GlobalSignalsHandler` shows a professional understanding of GObject-based development.
> *   **Memory Management:** One of the biggest issues with Cinnamon extensions is memory leaks. This author explicitly manages `run_dispose()` on pixbufs (critical for image processing) and uses a robust signal handler to ensure all connections are disconnected when the extension is disabled.
> *   **Performance Optimization:** 
>     *   The `BlurTemplateManager` uses an **LRU (Least Recently Used) cache** for CSS strings to avoid re-generating complex CSS on every UI update.
>     *   Wallpaper analysis uses **pixel sampling (grid steps)** rather than checking every single pixel, which prevents the CPU from spiking during wallpaper changes.
> *   **Theming Logic:** The `ThemeUtils` and `ThemeDetector` classes are impressive. They implement **HSP (Highly Sensitive Poo) perceived brightness** and **WCAG contrast ratio** calculations. This ensures that when the extension "auto-extracts" colors, the text remains readable.
> *   **User Experience:** It provides a "Glow" system that replaces hardcoded borders, giving it a modern, "glassmorphism" look (macOS/Windows 11 style) that is currently missing from the default Cinnamon experience.
> 
> #### **Cons (The Risks)**
> *   **Heavy Use of Monkey Patching:** To style the OSD (Volume/Brightness overlays), Alt-Tab switcher, and Notifications, the extension "monkey patches" (overwrites) core Cinnamon methods (e.g., `AppSwitcher3D.prototype._init`).
>     *   *Risk:* If the Cinnamon team updates the internal code of the `appSwitcher` in a future version (e.g., Cinnamon 6.4 or 7.0), this extension could cause the shell to crash or the Alt-Tab menu to disappear.
> *   **Maintenance Burden:** Because it touches so many parts of the UI (Nemo, Alt-Tab, Desklets, Panels, OSDs), it is prone to "bit rot." It will require an active maintainer to fix it every time Cinnamon changes its internal JS API.
> 
> ---
> 
> ### 2. Is it suitable for the official `cinnamon-spices` repo?
> 
> The `linuxmint/cinnamon-spices-extensions` repository is the official "3rd party" hub. For a PR to be accepted there, it needs to follow specific standards.
> 
> **My Verdict:** **Yes, it should be merged**, but with specific caveats.
> 
> #### **Why it fits:**
> 1.  **Code Standards:** The code is cleaner and more "modern" than 80% of the extensions currently in the Spices repository. It uses ES6 classes and modular `require` statements properly.
> 2.  **Feature Demand:** "Blur" and "Transparency" are the most requested visual features for Cinnamon. This extension provides a "one-stop-shop" for these features, replacing 3-4 smaller, older extensions.
> 3.  **Safety:** The author has included "idempotent restore" logic. Even though they monkey patch, they check if they still "own" the patch before restoring it, which prevents conflicts with other extensions.
> 
> #### **Potential Roadblocks for the Mint Team:**
> *   **Review Time:** The sheer volume of code (17+ files) is a lot for the Mint team to audit.
> *   **The "Nemo" integration:** The extension attempts to style Nemo (the file manager) context menus. Since Nemo is a C-based app and the shell is JS, this type of integration can sometimes be finicky.
> 
> ---
> 
> ### 3. Suggestions for the PR Author
> If you are the one submitting this or communicating with the author, here are a few suggestions to increase the chances of it being accepted by the Linux Mint maintainers:
> 
> 1.  **Strict Linting:** Ensure the code passes `eslint` with the Cinnamon configuration.
> 2.  **Cinnamon Version Guarding:** Ensure the `metadata.json` strictly defines the `cinnamon-version`. The backdrop-filter CSS used requires relatively modern versions of the Cinnamon/Muffin compositor.
> 3.  **Monkey Patch Documentation:** Add comments in `osdStyler.js` and `alttabStyler.js` explaining *why* the patch is necessary (usually because Cinnamon lacks a native "style-changed" signal for those specific components).
> 4.  **Localisation:** The `settings-schema.json` is very detailed. Ensure that all strings are wrapped in the translation function (which the author already seems to have started in `extension.js`).
> 
> ### Final Summary
> This is **top-tier work**. It brings a modern aesthetic to Cinnamon that usually requires a lot of manual CSS hacking. As long as the author is committed to maintaining it as Cinnamon evolves, it would be a "Crown Jewel" extension for the Cinnamon Spices ecosystem.

I think the extension should be versioned, i.e. put into a "6.6" subfolder so that it won't be loaded on cinnamon versions is hasn't been tested on? Apart from that this looks OK to me.

rcalixte commented:
> Here's what Gemini thinks.

Can I politely ask that we not do this? It's low-value noise and in general, I would like to discourage the habit. Let's keep this a platform where we interact with each other and not each other's LLMs.

fredcw commented:
@rcalixte

>It's low-value noise

I completely disagree. My comment is for informational purposes only. LLMs are invaluable in finding issues and problems and will only become more so in the future. It won't be long before LLM generated code will be far more reliable than human written code.

>where we interact with each other and not each other's LLMs.

How am I not interacting? I looked at the code and It looks safe to me and safely undoes all it's changes when disabled. As long as it's versioned, I think it's fine, which is what I stated:

>I think the extension should be versioned, i.e. put into a "6.6" subfolder so that it won't be loaded on cinnamon versions is hasn't been tested on? Apart from that this looks OK to me.

There is no LLM interaction here as far as I can see.

drdrummie commented:
Hi all!
I'm working more to fix some subtle bugs in the new version. Yeah, monkey
patching is "fragile" and would be thankful if you could point me to some
examples if it could be made differently.
I do have some dead code from previous approaches and lint errors,
really try to clean them up. I'm aware from testing that I can't satisfy
every detail and be compatible with everything the community has already
made.
Wanted to do something different and give alternatives to default mint
styles. Nemo integration is something that is a "placeholder" for the
future, based on my research there is no way to get into the nemo process,
still I have some ideas and leave it here for the future. I can remove it
if needed.

I believe in a few days the new version should be ready, so I agree to put
it in draft until the new version.

Thanks!

pon, 4. svi 2026. u 16:42 Rick Calixte ***@***.***> napisao
je:

> *rcalixte* left a comment (linuxmint/cinnamon-spices-extensions#924)
> <https://github.com/linuxmint/cinnamon-spices-extensions/pull/924#issuecomment-4371967223>
>
> Here's what Gemini thinks.
>
> Can I politely ask that we not do this? It's low-value noise and in
> general, I would like to discourage the habit. Let's keep this a platform
> where we interact with each other and not each other's LLMs.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/linuxmint/cinnamon-spices-extensions/pull/924#issuecomment-4371967223>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BV2KCXGH6NM5N4KLONIUMY34ZCT55AVCNFSM6AAAAACF3KKLWKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DGNZRHE3DOMRSGM>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drdrummie commented:
Just to explain why is there so many things:
After @fredcw first review I was little down knowing that he is right about functionality..
Needed to regroup and made [CSSGnome](https://github.com/drdrummie/CSSGnomme-ZorinOS18) for Zorin, so I started to rewrite CSSPanels similar to CSSGnomme approach. Found that it does work but was really tricky. Gave up again and tried to combine both.

mtwebster commented on csspanels@dr.drummie/files/csspanels@dr.drummie/constants.js:
```diff
@@ -0,0 +1,537 @@
+ */
+const SETTINGS_KEYS = {
+    // Basic transparency
+    PANEL_OPACITY: "panel-opacity",
+    MENU_OPACITY: "menu-opacity",
+    BORDER_RADIUS: "border-radius",
+    AUTO_DETECT_RADIUS: "auto-detect-radius",
+    APPLY_PANEL_RADIUS: "apply-panel-radius",
+
+    // Color overrides
+    OVERRIDE_PANEL_COLOR: "override-panel-color",
+    CHOOSE_OVERRIDE_PANEL_COLOR: "choose-override-panel-color",
+    OVERRIDE_POPUP_COLOR: "override-popup-color",
+    CHOOSE_OVERRIDE_POPUP_COLOR: "choose-override-popup-color",
+
+    // Feature toggles
+    ENABLE_NOTIFICATION_STYLING: "enable-notification-styling",
+    ENABLE_OSD_STYLING: "enable-osd-styling",
+    ENABLE_TOOLTIP_STYLING: "enable-tooltip-styling",
+    ENABLE_ALTTAB_STYLING: "enable-alttab-styling",
+    ENABLE_APPMENU_SIDEBAR_STYLING: "enable-appmenu-sidebar-styling",
+    ENABLE_DESKTOP_CONTEXT_STYLING: "enable-desktop-context-styling",
+
+    // System
+    HIDE_TRAY_ICON: "hide-tray-icon",
+    DEBUG_LOGGING: "debug-logging",
+    DARK_LIGHT_OVERRIDE: "dark-light-override",
+
+    // Blur settings
+    BLUR_RADIUS: "blur-radius",
+    BLUR_SATURATE: "blur-saturate",
+    BLUR_CONTRAST: "blur-contrast",
+    BLUR_BRIGHTNESS: "blur-brightness",
+    BLUR_BACKGROUND: "blur-background",
+    BLUR_BORDER_COLOR: "blur-border-color",
+    BLUR_BORDER_WIDTH: "blur-border-width",
+    BLUR_TRANSITION: "blur-transition",
+    BLUR_OPACITY: "blur-opacity",
+    BLUR_TEMPLATE: "blur-template",
```
These are all unused.

mtwebster commented:
Is this a fork of a gnome-shell extension?  How does nemoPopupStyler.js work? Nemo is a Gtk application, and unaffected by any CSS in Cinnamon.

drdrummie commented:
@mtwebster Yip I know, I believe I've answered that in post(s) above yours..

mtwebster commented:
We're going to pass on this.

The only thing guaranteed to happen when I enable this extension is that I'll have an unusable desktop.

There are lots of theme warnings in .xsession-errors. I see a lot of unused code. This wastes peoples' time trying to review this. I get you're trying to re-use an existing project but it should still be kept as tidy as possible.

You're trying to do too much here, honestly.  It's never going to work for all desktop elements, all third-party spices. Even with the native window-list applet, you can't touch the 'active' window, so it looks out of place from anything I set on the panel itself. You almost need a theme *that is designed* to be used with the extension - sort of a 'green screen' model that your customizations might be applied to - I don't know... but as it stands, I think someone will try this for a bit, get frustrated, then disable it.

drdrummie commented:
Ok, thank you for your time!
